### PR TITLE
Feature/randomness

### LINKED ISSUE #6: Feature request: Randomness
At the moment, there are no way to randomize the initial state of the NEST network with BSB. 
NEST has some random function implemented that could be used to randomize the values of certain parameters (e.g. V_m for neurons).
There should be a way to set the seed of the NEST simulation too.

### PR DESCRIPTION
**Pull request by drodarie**
- Add randomness to NEST simulations (seed) 
- Allow NEST random distribution for neuron parameters

close #6

### CODE REVIEW

File: bsb_nest/distributions.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import builtins
+import typing
+
+import errr
+import nest.random.hl_api_random as _distributions
+from bsb import DistributionCastError, TypeHandler, config, types
```
Comment (Helveg) [inline bsb_nest/distributions.py:6]: Did you add `DistributionCastError`? Has this been released already?

File: bsb_nest/distributions.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        except Exception as e:
+            errr.wrap(
+                DistributionCastError, e, prepend=f"Can't cast to '{self.distribution}': "
+            )
+
+    def __call__(self):
+        return self._distr
+
+    def __getattr__(self, attr):
+        if "_distr" not in self.__dict__:
```
Comment (Helveg) [inline bsb_nest/distributions.py:41]:
```suggestion
        if not hasattr(self, "_distr"):
```

File: bsb_nest/distributions.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    """
+
+    def __call__(self, value, _key=None, _parent=None):
+        if isinstance(value, builtins.dict) and "distribution" in value.keys():
+            return NestRandomDistribution(**value, _key=_key, _parent=_parent)
+        return value
+
+    @property
+    def __name__(self):  # pragma: nocover
+        return "nest_parameter"
```
Comment (Helveg) [inline bsb_nest/distributions.py]:
```suggestion
        return "nest parameter"
```

File: tests/test_nest.py [hunk 4]
```diff
@@ -210,3 +211,90 @@ def test_iaf_cond_alpha(self):
...
+                "mean": -70,
+                "std": 20.0,
+            },
+        )
+        # Test with an unknown distribution
+        conf["simulations"]["test"]["cell_models"]["A"]["constants"]["V_m"][
+            "distribution"
+        ] = "bean"
+        with self.assertRaises(CastError):
+            _ = Configuration(conf)
```
Comment (Helveg) [inline tests/test_nest.py]:
```suggestion
            Configuration(conf)
```

**Review by Helveg — CHANGES_REQUESTED** [review]

File: bsb_nest/distributions.py (hunk 1 above)
Comment (drodarie) [reply bsb_nest/distributions.py:6]: It was already there. It is also used for distribution attributes in bsb-core. I basically adapted the `Distribution` in bsb-core/bsb/config/_distributions.py

**Comment by drodarie** [comment]
Hmm for some reason the `hasattr` you suggested does not work. 
To give some context, I copied the implementation of Distribution in _distributions.py in bsb.core, and here too the `__getattr__` uses `self.__dict__`. 
I will revert the change and put a comment but I am not sure why this happens

**Comment by Helveg** [comment]
ah oops, I did not see that this was inside of `__getattr__`. It doesn't work because `hasattr` uses the entire Python attribute access cascade, including a lookup of `__getattr__` itself :)

**Review by Helveg — APPROVED** [review]
I'll approve it, but once `bsb-core` implements reproducibility/randomness systematically, this logic will probably have to be redone.

**Pull request merged**